File: spec/azk/box_spec.lua
local fs     = require('azk.utils.fs')
local path   = require('azk.utils.path')
local box    = require('azk.box')
local helper = require('spec.spec_helper')

describe("Azk box", function()
  it("should support github format", function()
    local result = box.parse("azukiapp/ruby-box#stable")

    assert.is.equal("github", result.type)
    assert.is.equal("https://github.com/azukiapp/ruby-box", result.repository)
    assert.is.equal("azukiapp/ruby-box", result.path)
    assert.is.equal("stable", result.version)
    assert.is.equal("azukiapp/ruby-box", result.image)
    assert.is.equal("azukiapp/ruby-box:stable", result.full_name)
  end)

  it("should support github format without version", function()
    local result = box.parse("azukiapp/ruby-box")

    assert.is.equal("github", result.type)
    assert.is.equal("master", result.version)
    assert.is.equal("azukiapp/ruby-box", result.image)
    assert.is.equal("azukiapp/ruby-box:master", result.full_name)
  end)

  it("should support path format", function()
    local box_path = helper.fixture_path("test-box") .. "/"
    local sha      = fs.shasum(box_path)
    local result   = box.parse(box_path)

    box_path = box_path:gsub("/$", ""):gsub("^/", "")

    assert.is.equal("path", result.type)
    assert.is.equal(nil, result.repository)
    assert.is.equal("/" .. box_path, result.path)
    assert.is.equal(sha, result.version)
    assert.is.equal(box_path, result.image)
    assert.is.equal(box_path .. ":" .. sha, result.full_name)
  end)

  it("should return a error if path not found", function()
    local box_path = path.join(fs.pwd(), "novalid")
    local _, err = pcall(box.parse, './novalid')
    local msg = helper.escape_regexp(
      "box directory '" .. box_path .. "' not found"
    )
    assert.is.match(err, msg)
  end)

  it("should expand relative path", function()
    fs.cd(helper.tmp_dir(), function()
      local box_path = "./box"
      fs.mkdir(box_path)

      local result = box.parse(box_path)

      assert.is.equal("path", result.type)
      assert.is.equal(path.join(fs.pwd(), "box"), result.path)
    end)
  end)

  it("should support docker format", function()
    local result = box.parse("ubuntu:12.04")

    assert.is.equal("docker", result.type)
    assert.is.equal(nil, result.repository)
    assert.is.equal(nil, result.path)
    assert.is.equal("12.04", result.version)
    assert.is.equal("ubuntu", result.image)
    assert.is.equal("ubuntu:12.04", result.full_name)
  end)

  it("should return erro for invalid box", function()
    local box_name = '%%#^%@'
    local _, err   = pcall(box.parse, box_name)
    local msg      = "'%s' is not a valid definition of box"

    assert.is.match(err, helper.escape_regexp(msg:format(box_name)))
  end)
end)

File: test/app/box_spec.js
var azk    = require('../../lib/azk');
var Box    = require('../../lib/app/box');
var sha1   = require('../../lib/utils/sha1');
var h      = require('../spec_helper');
var path   = require('path');
var Q      = require('q');

describe("Azk app box", function() {
  it("should support github format", function() {
    var box = new Box("azukiapp/ruby-box#stable");
    h.expect(box).to.have.property("type", "github");
    h.expect(box).to.have.property("origin", "https://github.com/azukiapp/ruby-box");
    h.expect(box).to.have.property("path", "azukiapp/ruby-box");
    h.expect(box).to.have.property("version", "stable");
    h.expect(box).to.have.property("repository", "azukiapp/ruby-box");
    h.expect(box).to.have.property("image", "azukiapp/ruby-box:stable");
  });

  it("should support github format without version", function() {
    var box = new Box("azukiapp/ruby-box");
    h.expect(box).to.have.property("type", "github");
    h.expect(box).to.have.property("origin", "https://github.com/azukiapp/ruby-box");
    h.expect(box).to.have.property("path", "azukiapp/ruby-box");
    h.expect(box).to.have.property("version", "master");
    h.expect(box).to.have.property("repository", "azukiapp/ruby-box");
    h.expect(box).to.have.property("image", "azukiapp/ruby-box:master");
  });

  it("should support path format", function() {
    return h.mock_app().then(function(box_path) {
      var repo     = box_path.replace(/^\//, '').replace(/\/$/, "")
      var hash     = sha1.calculateSync(box_path);

      var box = new Box(box_path);
      h.expect(box).to.have.property("type", "path");
      h.expect(box).to.have.property("origin", null);
      h.expect(box).to.have.property("path", box_path);
      h.expect(box).to.have.property("version", hash);
      h.expect(box).to.have.property("repository", repo);
      h.expect(box).to.have.property("image", repo + ":" + hash);
    });
  });

  it("should return a error if path not found", function() {
    var box_path = path.join(process.cwd(), "novalid")
    var func = function() { new Box(box_path); }
    h.expect(func).to.throw(Error, /box directory.*not found/);
  })

  it("should expand relative path", function() {
    var box_path = h.fixture_path("test-box")
    var repo     = box_path.replace(/^\//, '').replace(/\/$/, "")
    var hash     = sha1.calculateSync(box_path);

    var box = new Box(
      "./" + path.relative(process.cwd(), box_path)
    );

    h.expect(box).to.have.property("type", "path");
    h.expect(box).to.have.property("origin", null);
    h.expect(box).to.have.property("path", box_path);
    h.expect(box).to.have.property("version", hash);
    h.expect(box).to.have.property("repository", repo);
    h.expect(box).to.have.property("image", repo + ":" + hash);
  })

  it("should support docker format", function() {
    var image = azk.cst.DOCKER_DEFAULT_IMG.split(":");
    var box = new Box(azk.cst.DOCKER_DEFAULT_IMG);
    h.expect(box).to.have.property("type", "docker");
    h.expect(box).to.have.property("origin", null);
    h.expect(box).to.have.property("path", null);
    h.expect(box).to.have.property("version", image[1]);
    h.expect(box).to.have.property("repository", image[0]);
    h.expect(box).to.have.property("image", azk.cst.DOCKER_DEFAULT_IMG);
  });

  it("should return erro for invalid box", function() {
    var func = function() { new Box("%%#^%@"); }
    h.expect(func).to.throw(Error, /not a valid/);
  });
});

